working/0546-patterns/patterns.md

```diff
@@ -0,0 +1,923 @@
+
+```dart
+printGeometry(Geometry geometry) {
+  if (geometry is Point) {
+    print("Point ${geometry.y}, ${geometry.y}");
+  } else if (geometry is Circle) {
+    print("Circle ${geometry.x}, ${geometry.y}, ${geometry.radius}")
+  }
+}
+```
+
+Here we rely on type promotion on `geometry` to let us access the fields after
+checking the variable's type. With pattern matching we can combine that type
+test and field access like so:
+
+```dart
+printGeometry(Geometry geometry) {
+  match (geometry) {
+    case Point(var x, var y) => print("Point $y, $y");
+    case Circle(var x, var y, var radius) => print("Circle $x, $y, $radius");
+  }
+}
+```
+
+Here, the outer `Point(...)` pattern is a matcher that does a type test to see
+if `geometry` is a Point. If it is not, it simply proceeds to the next case. If
+`geometry` is a point, the `Point(...)` patterns destructures the value and
+evaluates its two subpatterns. The `var x` and `var y` patterns are *bind
+matchers* that hop over to the bind category to bind variables. A bind matcher
+is a special kind of matcher that contains a binder as a subpattern. This lets
+you compose binders and matchers together, much like expression statements let
+you place an expression where a statement is needed.
+
+Note that you can't cross categories in the other direction. In a binder, there
+is no way to get to a matcher. That makes sense because in a context where
+failure isn't meaningful and the user does not expect control flow, there should
+be no way for a match to fail.
+
+## Matchers
+
+Matchers determine if the value in question *matches* or meets some predicate.
+This answer is used to select appropriate control flow in the surrounding
+construct. For example, in some hypothetical match statement, if the pattern
+does not match, then the case containing the pattern is skipped.
+
+```
+matcher ::=
+  | wildcardMatcher
+  | literalMatcher
+  | constantMatcher
+  | listMatcher
+  | mapMatcher
+  | typedMatcher
+  | bindMatcher
+  | typeTestMatcher
+  | tupleMatcher
+  | extractMatcher
+
+matchers ::= matcher ( "," matcher )* ","?
+```
+
+**TODO: Do set matchers make sense?**
+
+### Wildcard matcher
+
+```
+wildcardMatcher ::= "_"
+```
+
+A wildcard pattern always matches. This is useful in places where a subpattern
+is required but you always want to succeed. It can function as a "default"
+pattern for the last case in a pattern matching statement.
+
+### Literal matcher
+
+```
+literalMatcher ::=
+  | booleanLiteral
+  | numericLiteral
+  | stringLiteral
+```
+
+A literal pattern determines if the value is equivalent to the given literal
+value. The pattern matches if the two values are equal, determined by passing
+the pattern's literal value to the `==` method on the value being matched.
+
+### Constant matcher
+
+```
+constantMatcher ::= identifier ( "." identifier ( "." identifier )? )?
+```
+
+Determines if the value is equivalent to the value of the given constant
```

Comment:
Matching against non-constant values could be useful, but I don't think most other languages support it and I think it would interact poorly with static exhaustiveness checking. We could conceivably support it but I don't know if I feel that it's well-motivated. We don't support non-constant values in switch statements, and that hasn't been much a limitation.